File: Loenn/entities/TimedDashSwitch.lua
local drawableSprite = require("structs.drawable_sprite")

local TimedDashSwitch = {}

TimedDashSwitch.name = "XaphanHelper/TimedDashSwitch"
TimedDashSwitch.depth = 0
TimedDashSwitch.fieldOrder = {
    "x", "y", "side", "spriteName", "timer", "mode", "tickingType", "flag", "inWall", "particleColor1", "particleColor2"
}
TimedDashSwitch.fieldInformation = {
    side = {
        options = {"Left", "Right", "Up", "Down"},
        editable = false
    },
    timer = {
        fieldType = "integer",
    },
    mode = {
        options = {"add", "set"},
        editable = false
    },
    tickingType = {
        options = {"On Top", "Tick only", "None"},
        editable = false
    },
    particleColor1 = {
        fieldType = "color"
    },
    particleColor2 = {
        fieldType = "color"
    }
}
TimedDashSwitch.placements = {
    name = "TimedDashSwitch",
    data = {
        side = "Up",
        spriteName = "default",
        timer = 10,
        mode = "add",
        tickingType = "Tick only",
        particleColor1 = "99E550",
        particleColor2 = "D9FFB5",
        flag = "",
        inWall = false
    }
}

function TimedDashSwitch.sprite(room, entity)
    local sprites = {}
    sprites["default"] = "objects/temple/dashButton00"
    sprites["mirror"] = "objects/temple/dashButtonMirror00"
    local texture = entity.spriteName == "default" and sprites["default"] or sprites["mirror"]
    local sprite = drawableSprite.fromTexture(texture, entity)
    local side = entity.side
    local inWall = entity.inWall or false

    if side == "Down" then
        sprite:addPosition(8, inWall and -4 or 0)
        sprite.rotation = -math.pi / 2
    elseif side == "Up" then
        sprite:addPosition(8, inWall and 12 or 8)
        sprite.rotation = math.pi / 2
    elseif side == "Left" then
        sprite:addPosition(inWall and -4 or 0, 8)
        sprite.rotation = math.pi
    elseif side == "Right" then
        sprite:addPosition(inWall and 12 or 8, 8)
        sprite.rotation = 0
    end

    return sprite
end

return TimedDashSwitch

File: Ahorn/entities/TimedDashSwitch.jl
module XaphanHelperTimedDashSwitch

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/TimedDashSwitch" TimedDashSwitch(x::Integer, y::Integer, width::Integer=16, height::Integer=8, side::String="Up", spriteName::String="default", timer::Integer=10, mode::String="add", tickingType::String="Tick only", particleColor1::String="99E550", particleColor2::String="D9FFB5", flag::String="", inWall::Bool=false)

const placements = Ahorn.PlacementDict(
    "Timed Dash Switch (Xaphan Helper)" => Ahorn.EntityPlacement(
        TimedDashSwitch,
        "point",
        ),
    )

function Ahorn.selection(entity::TimedDashSwitch)
    x, y = Ahorn.position(entity)
    side = get(entity.data, "side", "Up")
    inWall = get(entity.data, "inWall", false)
    if side == "Left"
        return Ahorn.Rectangle(x - (inWall ? 4 : 0), y, 10, 16)
    end
    if side == "Right"
        return Ahorn.Rectangle(x - 2 + (inWall ? 4 : 0), y, 10, 16)
    end
    if side == "Up"
        return Ahorn.Rectangle(x, y - 4 + (inWall ? 4 : 0), 16, 12)
    end
    if side == "Down"
        return Ahorn.Rectangle(x, y - (inWall ? 4 : 0), 16, 12)
    end
end

Ahorn.editingOptions(entity::TimedDashSwitch) = Dict{String, Any}(
    "side" => String["Left", "Right", "Up", "Down"],
    "mode" => String["Add", "Set"],
    "tickingType" => String["On top", "Tick only", "None"]
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TimedDashSwitch, room::Maple.Room)
    sprite = get(entity.data, "spriteName", "default")
    side = get(entity.data, "side", "Up")
    inWall = get(entity.data, "inWall", false)
    texture = sprite == "default" ? "objects/temple/dashButton00.png" : "objects/temple/dashButtonMirror00.png"
    if side == "Left"
        Ahorn.drawSprite(ctx, texture, inWall ? 16 : 20, 26, rot=pi)
    end
    if side == "Right"
        Ahorn.drawSprite(ctx, texture, inWall ? 12 : 8, 8)
    end
    if side == "Up"
        Ahorn.drawSprite(ctx, texture, 27, inWall ? 11 : 7, rot=pi / 2)
    end
    if side == "Down"
        Ahorn.drawSprite(ctx, texture, 9, inWall ? 16 : 20, rot=-pi / 2)
    end
end

end